cla_backend/apps/knowledgebase/tests/test_csv_user_import.py

```diff
@@ -0,0 +1,199 @@
+
+    def test_article_category_matrices_from_row__invalid_preferred_signpost(self):
+        make_recipe("knowledgebase.article_category", name="Employment")
+        article = make_recipe("knowledgebase.article")
+        with self.assertRaisesMessage(ValueError, "Yes is an invalid value for Preferred signpost"):
+            signpost_index = ARTICLE_CATEGORY_MATRIX_COLUMN_FIELD_MAPPING[0][1][0]
+            self.row[signpost_index] = "Yes"
+            self.importer.get_article_category_matrices_from_row(article, self.row)
+
+    def test_save(self):
+        self.assertEqual(Article.objects.count(), 0)
+        self.assertEqual(TelephoneNumber.objects.count(), 0)
+        self.assertEqual(ArticleCategoryMatrix.objects.count(), 0)
+
+        make_recipe("knowledgebase.article_category", name="Employment")
+        make_recipe("knowledgebase.article_category", name="Discrimination")
+
+        article = self.importer.get_article_from_row(self.row)
+        telephone_numbers = self.importer.get_telephone_numbers_from_row(article, self.row)
+        article_category_matrices = self.importer.get_article_category_matrices_from_row(article, self.row)
+        rows = [
+            {
+                "article": article,
+                "telephone_numbers": telephone_numbers,
+                "article_category_matrices": article_category_matrices,
+            }
+        ]
+
+        self.importer.save(rows)
+
+        self.assertEqual(Article.objects.count(), 1)
+        self.assertEqual(article.article_category.count(), 2)
+        self.assertEqual(TelephoneNumber.objects.filter(article=article).count(), 2)
+
+    def test_save_transaction(self):
+        # We are not going to create the categories that the ArticleCategoryMatrix.article_category requires
+        # This should cause an error when saving the ArticleCategoryMatrix object
+
+        self.assertEqual(Article.objects.count(), 0)
+        self.assertEqual(TelephoneNumber.objects.count(), 0)
+        self.assertEqual(ArticleCategoryMatrix.objects.count(), 0)
+
+        make_recipe("knowledgebase.article_category", name="Employment")
+        make_recipe("knowledgebase.article_category", name="Discrimination")
+
+        article = self.importer.get_article_from_row(self.row)
+        telephone_numbers = self.importer.get_telephone_numbers_from_row(article, self.row)
+        article_category_matrices = self.importer.get_article_category_matrices_from_row(article, self.row)
+
+        rows = [
+            {
+                "article": article,
+                "telephone_numbers": telephone_numbers,
+                "article_category_matrices": article_category_matrices,
+            }
+        ]
+
+        def save_article_category_matrices(*args, **kwargs):
+            self.assertEqual(Article.objects.count(), 1)
+            self.assertEqual(TelephoneNumber.objects.count(), 2)
+            raise ValueError("Cannot save article category matrices")
```

Comment:
I think I understand why this is mocked, and I like that it allows us to assert the article and phone number counts mid-transaction, but is it possible to let the exception come from the original code at all? ie is it possible to call `super()`?

Reply:
I don't think the test `save_article_category_matrices` gets a self argument, so it can't seem to do super.

Instead I have made a copy of the original method and called that after doing asserts:
`        def save_article_category_matrices(article, matrices):
            self.assertEqual(Article.objects.count(), 1)
            self.assertEqual(TelephoneNumber.objects.count(), 2)
            matrices[0] = None
            return self.importer._original_save_article_category_matrices(article, matrices)

        self.importer._original_save_article_category_matrices = self.importer.save_article_category_matrices
        self.importer.save_article_category_matrices = save_article_category_matrices
`

Reply:
Ideally I would have like to have done `rows[0]["article_category_matrices"][0].article = None` inside the `with` context that is expecting the exception but it seems django is doing model not null checking at assignment

Reply:
Ah yes my mistake, and that looks like a really clever solution to me